ocaml/runtime/shared_heap.c

```diff
@@ -958,8 +1018,417 @@ static void compact_update_ephe_list(volatile value *ephe_p)
+          total_live_blocks++;
+        }
+        p += wh;
+      }
+
+      cur_pool = cur_pool->next;
+      k++;
+    }
+
+    int i = 0;
+    pool** sz_pools = (pool**)caml_stat_alloc_noexc(sizeof(pool*) * total_pools);
+
+    /* add all full pools to sz_pools */
+    cur_pool = heap->unswept_full_pools[sz_class];
+    while(cur_pool) {
+      CAMLassert(i < full_pools);
+
+      sz_pools[i] = cur_pool;
+
+      i++;
+      cur_pool = cur_pool->next;
+    }
+
+    /* add all avail pools to sz_pools */
+    cur_pool = heap->unswept_avail_pools[sz_class];
+    while(cur_pool) {
+      CAMLassert(i < total_pools);
+
+      sz_pools[i] = cur_pool;
+
+      i++;
+      cur_pool = cur_pool->next;
+    }
+
+    /* Now sort sz_pools so that the largest chunks are first */
+    qsort(sz_pools, total_pools, sizeof(pool*), compact_compare_pools);
+
+    /* Index of the first pool to evacuate is the ceiling of the total live
+       blocks divided by the number of blocks in a pool. */
+    int evac_idx = (total_live_blocks + pool_blocks - 1) / pool_blocks;
+
+    /* We can have zero pools for a size class and that's fine */
+    if( total_live_blocks > 0 ) {
+      atomic_fetch_add(&global_live_pools, evac_idx);
+    }
+
+    /* mark all pools from 0 to evac_idx as not evacuated */
+    for( i = 0 ; i < evac_idx && i < total_pools; i++ ) {
+      pool* p = sz_pools[i];
+
+      p->evacuate = 0;
+    }
+
+    /* mark all pools from evac_idx to total_pools as evacuating */
+    for( i = evac_idx; i < total_pools; i++ ) {
+      pool* p = sz_pools[i];
+
+      p->evacuate = 1;
+    }
+
+    caml_stat_free(sz_pools);
+  }
+}
+
+/* This is the second phase of compaction. This function should only
+   be called from a single domain (and sandwiched between barriers).
+   Returns 1 if it's worth us doing a second phase or 0 if not. 
+   
+   In this function a single domain will create a sorted array of all
+   pools in the process and then mark as evacuating all but the
+   highest sorted `global_live_pools` pools. */
+int compact_phase_two_mark(int participating_count,
+                            caml_domain_state** participants) {
+  
+  /* We need to get the full set of pools across all domains,
+     including the free list */
+  int pools_count = 0;
+  int free_pools = 0;
+
+  for( int i = 0; i < participating_count; i++ ) {
+    struct caml_heap_state* heap = participants[i]->shared_heap;
+
+    for( int sz = 1; sz < NUM_SIZECLASSES; sz++ ) {
+      /* Just recheck this */
+      CAMLassert(heap->full_pools[sz] == NULL);
+      CAMLassert(heap->avail_pools[sz] == NULL);
+
+      int full_pools_count = compact_count_pools(heap->unswept_full_pools[sz]);
+      int avail_pools_count = compact_count_pools(heap->unswept_avail_pools[sz]);
+
+      pools_count += full_pools_count + avail_pools_count;
+    }
+  }
+
+  /* We need to drain any free_pools that have already been batch allocated. In
```

Comment:
Here, `pools_count` is the same as `global_live_pools`, which is only actually used in this function. So we could just put `int live_pools = pools_count;' here and lose the global.

Reply:
Good call, eliminated it.